### ISSUE TITLE
Select component encountering 'TypeError: Cannot read property 'id' of undefined' in createRenderableFavorites

### ISSUE THREAD

**Issue by mhuth** [patternfly/patternfly-react#4890]
**Describe the issue. What is the expected and unexpected behavior?**
Getting this error when using the `<Select>` component:
```
instrument.js:110 TypeError: Cannot read property 'id' of undefined
    at util.js:298
    at Array.filter (<anonymous>)
    at Object.t.createRenderableFavorites (util.js:298)
```

**Please provide the steps to reproduce. Feel free to link CodeSandbox or another tool.**

In our application the `<Select>` component has a few child components in it, including an array of `<SelectGroup>` and `<SelectOption>`s.  However the array element doesn't have a direct 'props' property, eg:
```
<Select>
  <InputGroup key: 'null'>...</InputGroup>
  <Divider key: "inline-filter-divider" ... />
  [array of ...
    <SelectGroup>
      <SelectOption>
     ...
      </SelectOption>
     </SelectGroup>
  ]
  <Button key: "manage all tags">...</Button>
</Select>
```
This is the value of 'items' at the line causing the problem in createRenderableFavorites in helpers/utils.js:
```
e: Array(4)
0: {$$typeof: Symbol(react.element), key: null, ref: null, props: {…}, type: ƒ, …}
1: {$$typeof: Symbol(react.element), key: "inline-filter-divider", ref: null, props: {…}, type: ƒ, …}
2: [{…}]
3: {$$typeof: Symbol(react.element), key: "manage all tags", ref: null, props: {…}, type: ƒ, …}
```
... where you can see items[2] is an array without a 'props' element.

**Is this a bug or enhancement? If this issue is a bug, is this issue blocking you or is there a work-around?**
Its a bug and the workaround is to use an earlier version of react-core without the problematic code in helpers/utils.js.

Proposed fix for this issue is:
```
diff --git a/packages/react-core/src/helpers/util.ts b/packages/react-core/src/helpers/util.ts
index 106b83d..f476425 100644
--- a/packages/react-core/src/helpers/util.ts
+++ b/packages/react-core/src/helpers/util.ts
...
@@ -319,7 +320,7 @@ export const createRenderableFavorites = (
     return favoriteItems;
   }
   return (items as React.ReactElement[])
-    .filter(item => favorites.includes(item.props.id))
+    .filter(item => 'props' in item && favorites.includes(item.props.id))
     .map(item => React.cloneElement(item, { isFavorite: true, enterTriggersArrowDown: isEnterTriggersArrowDown }));
 };
```
However there are a number of lines containing `item.props.id` in that file, so it might be prudent to modify the other lines as well.

**What is your product and what release version are you targeting?**
Reproduced this in react-core 4.57.0

**Comment by KKoukiou** [comment]
Hey @mhuth.
If you pass the options of the Select component as a array containing the different React.Element s it would work as expected. So for your scenario something like:

```
  this.options = [
    <InputGroup>
      <TextArea name="textarea2" id="textarea2" aria-label="textarea with button" />
      <Button id="textAreaButton2" variant="control">
        Button
      </Button>
    </InputGroup>,
    <Divider />,
    <SelectGroup label="Status" key="group1">
      <SelectOption key={3} value="Degraded" />
      <SelectOption key={4} value="Needs Maintenence" />
    </SelectGroup>,
    <Divider />,
    <SelectGroup label="Vendor Names" key="group2">
      <SelectOption key={6} value="Samsung" isDisabled />
      <SelectOption key={7} value="Hewlett-Packard" />
    </SelectGroup>,
    <Button id="textAreaButton2" variant="control">
      Button
    </Button>
  ];

....

  <Select
    ...
    >
    {this.options}
  </Select>
```

**Comment by nicolethoen** [comment]
@mhuth are you using favorites?

the docs state that an `id` prop is required when using the favorites feature and the error is being thrown from within the `createRenderableFavorites` function in `utils.ts`. If you are not using favorites, then that implies that there may be a bug causing `createRenderableFavorites` to be called erroneously.

**Comment by mhuth** [comment]
@nicolethoen No we are not using favorites.  FWIW this is our code:

```
<Select
            className='dropDownOverride'
            variant={SelectVariant.checkbox}
...
        >
            <InputGroup className='tags-filter-group'>
...
            </InputGroup>
            <Divider key="inline-filter-divider"/>
            {data.map((group, index) =>
                <SelectGroup key={`group${index}`} label={group.source}>
                    {group.data.map((tag, tagIndex) =>
                        <span key={tagIndex} className='tags-select-group'>
                            <SelectOption value={tag.id} isChecked={
                                selectedTags.find(selection => selection === tag.id)
                            }>
                                <Tooltip content={`${decodeURIComponent(tag.id)}`} position={TooltipPosition.right}>
                                    <span>{`${decodeURIComponent(tag.key + '=' + tag.value)}`}</span>
                                </Tooltip>
                            </SelectOption>
                            <Badge className='tags-select-badge'> {tag.count} </Badge>
                        </span>
                    )}
                </SelectGroup>
            )}
            <Button key='manage all tags'
...
            </Button>
        </Select>}
```

And these are some of the values at the time of error:
```
class Select extends React.Component {
    constructor() {
...
            if (this.props.favorites.length !== prevProps.favorites.length ||   <==== both are 0
                this.state.typeaheadFilteredChildren !== prevState.typeaheadFilteredChildren) {
                const tempRenderableChildren = this.props.variant === 'typeahead' || this.props.variant === 'typeaheadmulti'  <==== both are 'checkbox'
                    ? this.state.typeaheadFilteredChildren
                    : this.props.children;
                const renderableFavorites = util_1.createRenderableFavorites(tempRenderableChildren, this.props.isGrouped, this.props.favorites);
...
```
`this.state.typeaheadFilteredChildren` is
```
0: {$$typeof: Symbol(react.element), key: ".0", ref: null, props: {…}, type: ƒ, …}
1: {$$typeof: Symbol(react.element), key: ".$inline-filter-divider", ref: null, props: {…}, type: ƒ, …}
2: {$$typeof: Symbol(react.element), key: ".2:$group0", ref: null, props: {…}, type: ƒ, …}
3: {$$typeof: Symbol(react.element), key: ".$manage all tags", ref: null, props: {…}, type: ƒ, …}
length: 4
__proto__: Array(0)
```

`this.props.children` is
```
0: {$$typeof: Symbol(react.element), key: null, ref: null, props: {…}, type: ƒ, …}
1: {$$typeof: Symbol(react.element), key: "inline-filter-divider", ref: null, props: {…}, type: ƒ, …}
2: Array(1)   <==== this is the problem because it doesn't have a direct 'props'
	0: {$$typeof: Symbol(react.element), key: "group0", ref: null, props: {…}, type: ƒ, …}
	length: 1
	__proto__: Array(0)
3: {$$typeof: Symbol(react.element), key: "manage all tags", ref: null, props: {…}, type: ƒ, …}
length: 4
__proto__: Array(0)
```
`this.props.isGrouped` is False
`this.props.favorites` is []

Hope this helps somehow.

**Comment by mhuth** [comment]
@nicolethoen @KKoukiou I should also mention our code was working fine on patternfly react-core v4.47 and stopped when upgrading to about v4.50 and later.  We have had to revert to using v4.47 for our app to run.  So the problem seems to be a bug introduced into pf react-core after that version.

**Comment by nicolethoen** [comment]
@mhuth 
I agree. It seems that there is an erroneous call to `createRenderableFavorites` in the `componentDidUpdate` function. It will be fixed in the next release! Thank you!

**Comment by tlabaj** [comment]
@mhuth this issue will be fixed in PR #4772

**Comment by KKoukiou** [comment]
#4890 is already merged which fixed this, closing this.
